src/LinAlg/hiopLinSolverSparseCUSOLVER.cpp

```diff
@@ -351,517 +433,1099 @@ namespace hiop
+    cudaFree(devr_);
+    devx_ = nullptr;
+    devr_ = nullptr;
+    checkCudaErrors(cudaMalloc(&devx_, n_ * sizeof(double)));
+    checkCudaErrors(cudaMalloc(&devr_, n_ * sizeof(double)));
 
-    cusparseCreateMatDescr(&descr_M_);
-    cusparseSetMatType(descr_M_, CUSPARSE_MATRIX_TYPE_GENERAL);
-    cusparseSetMatIndexBase(descr_M_, CUSPARSE_INDEX_BASE_ZERO);
+    cudaFree(dia_);
+    cudaFree(da_);
+    cudaFree(dja_);
+    /* this has to be done no matter what */
 
-    //info (data structure where factorization is stored)
-    cusolverSpCreateCsrluInfoHost(&info_lu_);
-    cusolverSpCreateGluInfo(&info_M_);
+    checkCudaErrors(cudaMalloc(&da_, nnz_ * sizeof(double)));
+    checkCudaErrors(cudaMalloc(&dja_, nnz_ * sizeof(int)));
+    checkCudaErrors(cudaMalloc(&dia_, (n_ + 1) * sizeof(int)));
+
+    checkCudaErrors(
+        cudaMemcpy(da_, kVal_, sizeof(double) * nnz_, cudaMemcpyHostToDevice));
+    checkCudaErrors(cudaMemcpy(dia_, kRowPtr_, sizeof(int) * (n_ + 1),
+                               cudaMemcpyHostToDevice));
+    checkCudaErrors(
+        cudaMemcpy(dja_, jCol_, sizeof(int) * nnz_, cudaMemcpyHostToDevice));
+    /*
+     * initialize KLU and cuSolver parameters
+     */
 
-    //KLU 
+    if(fact_ == "KLU") {
+      /* initialize KLU setup parameters, dont factorize yet */
+      this->initializeKLU();
 
-    klu_defaults(&Common_) ;
+      /*perform KLU but only the symbolic analysis (important)   */
 
-    // TODO: consider making a part of setup options that can be called from a user side
-    // For now, keeping these options hard-wired
+      klu_free_symbolic(&Symbolic_, &Common_);
+      klu_free_numeric(&Numeric_, &Common_);
+      Symbolic_ = klu_analyze(n_, kRowPtr_, jCol_, &Common_);
+
+      if(Symbolic_ == nullptr) {
+        nlp_->log->printf(hovError,
+                          "symbolic nullptr\n"); // catastrophic failure
+      }
+    } else { // for future
+      assert(0 && "Only KLU initial factorization is available.\n");
+    }
+  }
+
+  int
+  hiopLinSolverNonSymSparseCUSOLVER::initializeKLU()
+  {
+    klu_defaults(&Common_);
+
+    // TODO: consider making a part of setup options that can be called from a
+    // user side For now, keeping these options hard-wired
     Common_.btf = 0;
-    Common_.ordering = 1;//COLAMD; use 0 for AMD
-    Common_.tol = 0.1;
+    Common_.ordering = 1; // COLAMD; use 0 for AMD
+    Common_.tol = 0.01;
     Common_.scale = -1;
-    Common_.halt_if_singular=0;
+    Common_.halt_if_singular = 1;
 
-    free(mia_); 
-    free(mja_);
-    mia_ = nullptr;
-    mja_ = nullptr;
-    //allocate gpu data
-    cudaFree(devx_);
-    cudaFree(devr_);
-    devx_ = nullptr;
-    devr_ = nullptr;
-    checkCudaErrors(cudaMalloc(&devx_, n_ * sizeof(double)));
-    checkCudaErrors(cudaMalloc(&devr_, n_ * sizeof(double)));
-    this->newKLUfactorization();
+    return 0;
   }
 
+  int
+  hiopLinSolverNonSymSparseCUSOLVER::initializeCusolverGLU()
+  {
+
+    printf("CUSOLVER: Glu \n");
```

Comment:
is this supposed to be here?

Reply:
Good catch! This is supposed to write to HiOp logger.